### PR TITLE
Add agent liveness reporting to felix/neutron

### PR DESCRIPTION
**Pull request by fasaxc**
* Felix reports its liveness into etcd with a TTL.
* Neutron driver polls etcd for felix status and reports into neutron's DB.
* Neutron driver blocks VM scheduling if felix is not running.
* etcd polling function extracted out into etcdutils.EtcdWatcher class.

Hat tip to @miroslav21 who wrote the original version of this patch (300ee04e0121391d387629d5469668baf2f8a641).  I have reworked it to apply code review markups and cleaned up the history, which had become too difficult to rebase.

### CODE REVIEW

File: calico/datamodel_v1.py [hunk 1]
```diff
@@ -114,6 +133,37 @@ def get_profile_id_for_profile_dir(key):
...
+def hostname_from_status_key(key):
+    """
+    Get hostname from a status key (or None if this is not a status key).
+
+    :param: key for felix status
+            expected key format: FELIX_STATUS_DIR/<hostname>/
+                                           <some path or not>/<actual key name>
+    """
+    if not key.startswith(FELIX_STATUS_DIR):
+        return False
```
Comment (alexwlchan) [inline calico/datamodel_v1.py]: Docstring says a non-status key returns None, but it seems to return False.

File: calico/felix/test/test_config.py [hunk 2]
```diff
@@ -273,4 +285,117 @@ def test_ip_in_ip_enabled_bad(self):
...
+            config = Config("calico/felix/test/data/felix_missing.cfg")
+        cfg_dict = { "InterfacePrefix": "blah",
+                     "ReportingTTLSecs": "NaN"}
+        with self.assertRaisesRegexp(ConfigException,
+                                     "Field was not integer.*"):
+            config.report_etcd_config({}, cfg_dict)
+
+    def test_reporting_interval_not_int(self):
+        """
+        Test exception is rased ifstatus reporting interval has invalid (non-integer) value.
```
Comment (alexwlchan) [inline calico/felix/test/test_config.py]: Typo: "rased ifstatus"

File: calico/felix/test/test_config.py [hunk 3]
```diff
@@ -273,4 +285,117 @@ def test_ip_in_ip_enabled_bad(self):
...
+                     "ReportingTTLSecs": 47}
+        with mock.patch('calico.common.complete_logging'):
+            config.report_etcd_config({}, cfg_dict)
+
+        self.assertEqual(config.REPORTING_INTERVAL_SECS, 42)
+        self.assertEqual(config.REPORTING_TTL_SECS, 47)
+
+    def test_reporting_ttl_less_than_interval(self):
+        """
+        Test that exception is raised when status reporting ttl is less than interval.
```
Comment (alexwlchan) [inline calico/felix/test/test_config.py]: There's nothing in this test which looks for an exception, nor does the code it's testing raise an exception. It's just checking that a too-low reporting TTl is correctly defaulted.

File: calico/felix/test/test_config.py [hunk 4]
```diff
@@ -273,4 +285,117 @@ def test_ip_in_ip_enabled_bad(self):
...
+        cfg_dict = { "InterfacePrefix": "blah",
+                     "ReportingIntervalSecs": "21" }
+        with mock.patch('calico.common.complete_logging'):
+            config.report_etcd_config({}, cfg_dict)
+
+        self.assertEqual(config.REPORTING_TTL_SECS, 52)
+
+    def test_valid_interval_and_ttl(self):
+        """
+        Test valid reporting parameters are not chaned.
```
Comment (alexwlchan) [inline calico/felix/test/test_config.py]: Typo: "chaned"

File: calico/openstack/mech_calico.py [hunk 5]
```diff
@@ -1038,3 +1084,18 @@ def profiles_match(etcd_profile, neutron_profile):
         (etcd_rules == neutron_group_rules) and
         (etcd_tags == neutron_group_tags)
     )
+
+
+def felix_agent_state(hostname, start_flag=False):
+    """
+    :param bool start_flag: True if this is a new felix, that is starting up.
+           false if this is a refresh of an existing felix.
```
Comment (alexwlchan) [inline calico/openstack/mech_calico.py]: Nitpick: "false" ~> "False"

File: calico/openstack/test/test_plugin_etcd.py [hunk 6]
```diff
@@ -738,6 +767,159 @@ def test_not_master_does_not_resync(self):
...
+        self.driver.db = mock.Mock()
+        self.driver.transport = mock.Mock()
+        self.driver.transport.status_client = self.client
+        # Read through status subtree.
+        self.driver._register_initial_felixes()
+        self.driver.db.create_or_update_agent.assert_has_calls(expected_calls,
+                                                               any_order=True)
+
+        # Reset database
+        self.etcd_data = []
```
Comment (alexwlchan) [inline calico/openstack/test/test_plugin_etcd.py]: The etcd_data attribute is initialised to a dict, not a list. I'm also not convinced we need to reset at the end of a test; I think this isolation is enforced by nose.

File: calico/openstack/test/test_plugin_etcd.py [hunk 7]
```diff
@@ -764,3 +946,45 @@ def _neutron_rule_from_dict(overrides):
...
+            return False
+        if self.host != None and other["host"] != self.host:
+            return False
+        if self.topic != None and other["topic"] != self.topic:
+            return False
+        if self.start_flag and not other.get("start_flag"):
+            return False
+        if self.start_flag == False and other.get("start_flag"):
+            return False
+        return  True
```
Comment (alexwlchan) [inline calico/openstack/test/test_plugin_etcd.py]: Nitpick: extra space, and there should be a newline separating these two methods.

**Comment by fasaxc** [comment]
@alexwlchan Please can you review this PR, since you've got the most experience with it?  Would it be worth a quick chat to explain the larger changes that I made to Miroslav's code?

**Comment by alexwlchan** [comment]
@fasaxc Sounds good. I’m fairly busy this morning, but I’ll call you after lunch to discuss the changes.

File: calico/openstack/test/test_plugin_etcd.py (hunk 6 above)
Comment (fasaxc) [inline calico/openstack/test/test_plugin_etcd.py]: Actually, this test is completely broken but it doesn't appear to be running when I run the UTs locally!  I'll have a dig.

File: CHANGES.md [hunk 8]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## 1.2.0
+
+- Add liveness reporting to felix.  Felix now reports its liveness into
+  etcd and the neutron driver copies that information to the neutron DB.
```
Comment (alexwlchan) [inline CHANGES.md]: Tiny nitpick: "felix" and "neutron" should both be capitalised.

**Comment by matthewdupre** [comment]
ok to test

File: calico/felix/test/base.py [hunk 9]
```diff
@@ -1,4 +1,5 @@
 # Copyright (c) Metaswitch Networks 2015. All rights reserved.
```
Comment (alexwlchan) [inline calico/felix/test/base.py]: Does this file need a license?

File: calico/felix/test/base.py [hunk 10]
```diff
@@ -34,3 +35,28 @@ def step_actor(self, actor):
...
+        other_as_obj = None
+        try:
+            other_as_obj = json.loads(other)
+        except (ValueError, KeyError):
+            return False
+        if other_as_obj == self.json_obj:
+            return True
+        else:
+            _log.error("JSON didn't match %s != %s",
+                       self.json_obj, other_as_obj)
```
Comment (alexwlchan) [inline calico/felix/test/base.py]: Should we be returning False along with this log?

File: calico/felix/test/test_fetcd.py [hunk 11]
```diff
@@ -450,3 +461,79 @@ def dispatch(self, key, action, value=None):
...
+            self.assertRaises(RuntimeError,
+                              self.api._periodically_report_status)
+        self.assertEqual(m_update.mock_calls,
+                         [call(10)] * 3)
+
+        retry_call, jittered_call = m_sleep.mock_calls
+        self.assertEqual(retry_call, call(5))
+        _, (delay,), _ = jittered_call
+        self.assertTrue(delay >= 1)
+        self.assertTrue(delay <= 1.1005)
```
Comment (alexwlchan) [inline calico/felix/test/test_fetcd.py]: unittest provides the slightly more specific assertGreaterEqual and assertLessEqual; probably preferable.

File: calico/openstack/t_etcd.py [hunk 12]
```diff
@@ -527,6 +530,63 @@ def stop(self):
...
+        port = cfg.CONF.calico.etcd_port
+        LOG.info("CalicoEtcdWatcher created for %s:%s", host, port)
+        super(CalicoEtcdWatcher, self).__init__(
+            "%s:%s" % (host, port),
+            FELIX_STATUS_DIR
+        )
+        self.calico_driver = calico_driver
+
+        # Register for felix uptime updates.
+        self.register_path(FELIX_STATUS_DIR + "/<hostname>/status",
```
Comment (alexwlchan) [inline calico/openstack/t_etcd.py]: Is this supposed to be a real hostname, or is registering a placeholder path deliberate?

**Comment by alexwlchan** [comment]
A handful of further comments and nitpicks, but otherwise LGTM.

File: calico/openstack/t_etcd.py (hunk 12 above)
Comment (fasaxc) [inline calico/openstack/t_etcd.py]: `<foo>` is a capture, the path dispatcher captures the path segment there and passes it to the registered function as the `hostname=...` kwarg.

File: calico/felix/test/test_fetcd.py (hunk 11 above)
Comment (fasaxc) [inline calico/felix/test/test_fetcd.py]: Unless we have a monkey-patch in place those were only introduced in 2.7 :-(

File: calico/felix/test/base.py (hunk 10 above)
Comment (fasaxc) [inline calico/felix/test/base.py]: Yes.

File: calico/openstack/t_etcd.py (hunk 12 above)
Comment (alexwlchan) [inline calico/openstack/t_etcd.py]: Okay, makes sense. Just wanted to check.

File: calico/felix/test/test_fetcd.py (hunk 11 above)
Comment (alexwlchan) [inline calico/felix/test/test_fetcd.py]: Boo. :-(

**Comment by alexwlchan** [comment]
LGTM. @fasaxc Assigning back to you in case you want anybody else to review it; otherwise merge at your leisure.

**Pull request merged**